File: Flink/LinearRoad/src/main/java/LinearRoad/TollNotificationLas.java
package LinearRoad;

import org.apache.flink.api.common.functions.RichFlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.util.Collector;
import org.slf4j.Logger;
import common.LavTuple;
import common.SegmentIdentifier;
import common.TollNotification;
import util.Log;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

class TollNotificationLas extends RichFlatMapFunction<Tuple2<Long, LavTuple>, Tuple2<Long, TollNotification>> {
    private static final Logger LOG = Log.get(TollNotificationLas.class);

    //////////
    /**
     * Buffer for accidents.
     */
    private Set<SegmentIdentifier> currentMinuteAccidents = new HashSet<>();
    /**
     * Buffer for accidents.
     */
    private Set<SegmentIdentifier> previousMinuteAccidents = new HashSet<>();
    /**
     * Buffer for car counts.
     */
    private Map<SegmentIdentifier, Integer> currentMinuteCounts = new HashMap<>();
    /**
     * Buffer for car counts.
     */
    private Map<SegmentIdentifier, Integer> previousMinuteCounts = new HashMap<>();
    /**
     * Buffer for LAV values.
     */
    private Map<SegmentIdentifier, Integer> currentMinuteLavs = new HashMap<>();
    /**
     * Buffer for LAV values.
     */
    private Map<SegmentIdentifier, Integer> previousMinuteLavs = new HashMap<>();
    /**
     * The currently processed 'minute number'.
     */
    private int currentMinute = -1;
    //////////

    @Override
    public void flatMap(Tuple2<Long, LavTuple> tuple, Collector<Tuple2<Long, TollNotification>> collector) throws Exception {
        LOG.debug("TollNotificationLas");

        long timestamp = tuple.f0;

        collector.collect(new Tuple2<>(timestamp, new TollNotification()));

        LavTuple inputLavTuple = tuple.f1;

        this.checkMinute(inputLavTuple.minuteNumber);

        SegmentIdentifier segmentIdentifier = new SegmentIdentifier();
        segmentIdentifier.xway = inputLavTuple.xway;
        segmentIdentifier.segment = inputLavTuple.segment;
        segmentIdentifier.direction = inputLavTuple.direction;
        this.currentMinuteLavs.put(segmentIdentifier, inputLavTuple.lav);
    }

    private void checkMinute(short minute) {
        //due to the tuple may be send in reverse-order, it may happen that some tuples are processed too late.
//        assert (minute >= this.currentMinute);

        if (minute < this.currentMinute) {
            //restart..
            currentMinute = minute;
        }
        if (minute > this.currentMinute) {
            this.currentMinute = minute;
            this.previousMinuteAccidents = this.currentMinuteAccidents;
            this.currentMinuteAccidents = new HashSet<>();
            this.previousMinuteCounts = this.currentMinuteCounts;
            this.currentMinuteCounts = new HashMap<>();
            this.previousMinuteLavs = this.currentMinuteLavs;
            this.currentMinuteLavs = new HashMap<>();
        }
    }
}

File: WindFlow_3.x/LinearRoad/lib/linear_road/toll_notification_las.cpp
#include "toll_notification_las.hpp"
#include "util/log.hpp"

namespace linear_road {

void TollNotificationLas::operator ()(const LavTuple &tuple, wf::Shipper<NotificationTuple> &shipper, wf::RuntimeContext &rc)
{
    DEBUG_LOG("toll_notification_las::tuple" << tuple);

    NotificationTuple nt{};
    nt.ts = tuple.ts;
    shipper.push(std::move(nt)); // as an indication.

    check_minute(tuple.minute_number);

    SegmentIdentifier segment_identifier{tuple.xway, tuple.segment, tuple.direction};
    current_minute_lavs_.insert(std::make_pair(segment_identifier, tuple.lav));
}

void TollNotificationLas::check_minute(short minute)
{
    if (minute < current_minute_) {
        // restart...
        current_minute_ = minute;
    } else if (minute > current_minute_) {
        current_minute_ = minute;
        previous_minute_accidents_.swap(current_minute_accidents_);
        current_minute_accidents_.clear();
        previous_minute_counts_.swap(current_minute_counts_);
        current_minute_counts_.clear();
        previous_minute_lavs_.swap(current_minute_lavs_);
        current_minute_lavs_.clear();
    }
}

}
